app/controllers/submissions_controller.rb

```diff
@@ -8,28 +7,43 @@ def new
 
   def create
     task = API::Task.includes(:framework).find(params[:task_id]).first
-    submission = API::Submission.create(task_id: task.id)
-    task.update(status: 'in_progress')
 
-    ActiveStorage::Blob.create_after_upload!(
-      io: params[:upload],
-      filename: params[:upload].original_filename,
-      content_type: params[:upload].content_type,
-      metadata: { submission_id: submission.id }
-    )
+    submission = upload_file_submission(task, params[:upload])
 
     redirect_to task_submission_path(task_id: task.id, id: submission.id)
   end
 
   def show
     @task = API::Task.includes(:framework).find(params[:task_id]).first
-    @submission = API::Submission.includes(:files, :entries).find(params[:id]).first
+    @submission = API::Submission.includes(:entries).find(params[:id]).first
 
     render template_for_submission(@submission)
   end
 
   private
 
+  def upload_file_submission(task, upload)
+    submission = API::Submission.create(task_id: task.id)
+    submission_file = API::SubmissionFile.create(submission_id: submission.id)
+
+    blob = ActiveStorage::Blob.create_after_upload!(
+      io: upload,
+      filename: upload.original_filename,
+      content_type: upload.content_type,
+      metadata: { submission_id: submission.id, submission_file_id: submission_file.id }
+    )
+
+    blob_attributes = blob
+                      .attributes
+                      .slice('key', 'filename', 'content_type', 'byte_size', 'checksum')
+                      .merge(file_id: submission_file.id)
```

Comment:
Does this need a comment perhaps? It took me a moment to realise `file_id` is used to build the create URL, rather than getting passed as a parameter to the API endpoint.

Reply:
I've added a comment.